### PR TITLE
fix: regularly check whether the federated token has expired

### PR DESCRIPTION
**Pull request by wcy-fdu**
as title.

### CODE REVIEW

**Comment by wcy-fdu** [comment]
resolving conflict

File: src/azure/storage/credential.rs [hunk 1]
```diff
@@ -16,5 +16,36 @@ pub enum Credential {
     /// associated with the subscription that contains the storage account.
     ///
     /// ref: <https://docs.microsoft.com/rest/api/storageservices/authorize-with-azure-active-directory>
-    BearerToken(String),
+    BearerToken(String, String),
```
Comment (Xuanwo) [inline src/azure/storage/credential.rs:19]: It's better to store a `DateTime` directly. `is_valid` will check it every time we call `load_credential`.

File: src/azure/storage/loader.rs [hunk 2]
```diff
@@ -80,7 +84,10 @@ impl Loader {
         let workload_identity_token =
             workload_identity_credential::get_workload_identity_token(&self.config).await?;
         match workload_identity_token {
-            Some(token) => Ok(Some(Credential::BearerToken(token.access_token))),
+            Some(token) => Ok(Some(Credential::BearerToken(
+                token.access_token,
+                token.expires_on.unwrap_or("".to_string()),
```
Comment (Xuanwo) [inline src/azure/storage/loader.rs:89]: We can set to 600s later to enforce token refresh even when expires on not returned.

File: src/azure/storage/credential.rs [hunk 3]
```diff
@@ -26,11 +28,9 @@ impl Credential {
             return false;
         }
         if let Credential::BearerToken(_, expires_on) = self {
-            if let Ok(expires) = chrono::DateTime::parse_from_rfc3339(expires_on) {
-                let buffer = chrono::Duration::try_minutes(2).expect("in bounds");
-                if expires > (chrono::Utc::now() + buffer) {
-                    return false;
-                }
+            let buffer = chrono::TimeDelta::try_minutes(2).expect("in bounds");
+            if expires_on > &(chrono::Utc::now() + buffer) {
```
Comment (Xuanwo) [inline src/azure/storage/credential.rs:32]: Is this correct?

File: src/azure/storage/loader.rs [hunk 4]
```diff
@@ -72,10 +74,11 @@ impl Loader {
     async fn load_via_imds(&self) -> Result<Option<Credential>> {
         let token =
             imds_credential::get_access_token("https://storage.azure.com/", &self.config).await?;
-        let cred = Some(Credential::BearerToken(
-            token.access_token,
-            token.expires_on,
-        ));
+        let expires_on = match token.expires_on.is_empty() {
```
Comment (Xuanwo) [inline src/azure/storage/loader.rs:77]: match a `bool` doesn't look good, please use `if` directly.

Comment (wcy-fdu) [reply src/azure/storage/loader.rs:77]: I was wondering if I need to expire for IMDS credential, there was no such logic here before, but work well, right?

Comment (Xuanwo) [reply src/azure/storage/loader.rs:77]: > I was wondering if I need to expire for IMDS credential, there was no such logic here before, but work well, right?

It's better to have fun to enforce the refresh.

File: src/azure/storage/credential.rs (hunk 3 above)
Comment (wcy-fdu) [reply src/azure/storage/credential.rs:32]: Let's wait for the longevity test result.

**Review by Xuanwo — APPROVED** [review]
Thanks!

**Pull request merged**